modules/imgproc/src/pyramids.cpp

```diff
@@ -363,113 +284,63 @@ struct PyrUpVec_32s16s
-            __m128i v_dst10 = _mm_slli_epi32(_mm_add_epi32(v_r1, v_r2), 2);
-
-            v_r0 = _mm_loadu_si128((__m128i const *)(row0 + x + 4));
-            v_r1 = _mm_loadu_si128((__m128i const *)(row1 + x + 4));
-            v_r2 = _mm_loadu_si128((__m128i const *)(row2 + x + 4));
-            v_2r1 = _mm_slli_epi32(v_r1, 1);
-            v_4r1 = _mm_slli_epi32(v_r1, 2);
-            __m128i v_dst01 = _mm_add_epi32(_mm_add_epi32(v_r0, v_r2), _mm_add_epi32(v_2r1, v_4r1));
-            __m128i v_dst11 = _mm_slli_epi32(_mm_add_epi32(v_r1, v_r2), 2);
-
-            _mm_storeu_si128((__m128i *)(dst0 + x),
-                _mm_packs_epi32(_mm_srai_epi32(_mm_add_epi32(v_dst00, v_delta), 6),
-                                _mm_srai_epi32(_mm_add_epi32(v_dst01, v_delta), 6)));
-            _mm_storeu_si128((__m128i *)(dst1 + x),
-                _mm_packs_epi32(_mm_srai_epi32(_mm_add_epi32(v_dst10, v_delta), 6),
-                                _mm_srai_epi32(_mm_add_epi32(v_dst11, v_delta), 6)));
+            v_int32 v_r00 = vx_load(row0 + x),
+                    v_r01 = vx_load(row0 + x + v_int32::nlanes),
+                    v_r10 = vx_load(row1 + x),
+                    v_r11 = vx_load(row1 + x + v_int32::nlanes),
+                    v_r20 = vx_load(row2 + x),
+                    v_r21 = vx_load(row2 + x + v_int32::nlanes);
+            v_store(dst0 + x, v_rshr_pack<6>(v_r00 + v_r20 + ((v_r10 << 1) + (v_r10 << 2)), v_r01 + v_r21 + ((v_r11 << 1) + (v_r11 << 2))));
+            v_store(dst1 + x, v_rshr_pack<6>((v_r10 + v_r20) << 2, (v_r11 + v_r21) << 2));
         }
-
-        for( ; x <= width - 4; x += 4 )
+        if(x <= width - v_int32::nlanes)
         {
-            __m128i v_r0 = _mm_loadu_si128((__m128i const *)(row0 + x)),
-                    v_r1 = _mm_loadu_si128((__m128i const *)(row1 + x)),
-                    v_r2 = _mm_loadu_si128((__m128i const *)(row2 + x));
-            __m128i v_2r1 = _mm_slli_epi32(v_r1, 1), v_4r1 = _mm_slli_epi32(v_r1, 2);
-
-            __m128i v_dst0 = _mm_add_epi32(_mm_add_epi32(v_r0, v_r2), _mm_add_epi32(v_2r1, v_4r1));
-            __m128i v_dst1 = _mm_slli_epi32(_mm_add_epi32(v_r1, v_r2), 2);
-
-            _mm_storel_epi64((__m128i *)(dst0 + x),
-                _mm_packs_epi32(_mm_srai_epi32(_mm_add_epi32(v_dst0, v_delta), 6), v_zero));
-            _mm_storel_epi64((__m128i *)(dst1 + x),
-                _mm_packs_epi32(_mm_srai_epi32(_mm_add_epi32(v_dst1, v_delta), 6), v_zero));
+            v_int32 v_r00 = vx_load(row0 + x),
+                    v_r10 = vx_load(row1 + x),
+                    v_r20 = vx_load(row2 + x);
+            v_rshr_pack_store<6>(dst0 + x, v_r00 + v_r20 + ((v_r10 << 1) + (v_r10 << 2)));
+            v_rshr_pack_store<6>(dst1 + x, (v_r10 + v_r20) << 2);
+            x += v_int32::nlanes;
         }
 
         return x;
     }
 };
 
-#if CV_SSE4_1
+#if CV_SSE4_1 || CV_NEON
```

Comment:
Please, Could you add CV_VSX?

Reply:
@seiko2plus Could you please open a separate PR with update.

Reply:
@alalek , yeah, sure

Reply:
@seiko2plus, @alalek 
It make sense to remove this condition completely and make universal intrinsics code suitable for the common case. On my machine this code improves performance even on SSE2 baseline.

Reply:
@terfendail , Have you tried to run performance tests on unsigned 16-bit ? I mean the current benchmark that you attached missing this type, you may need to add it into Pyramid's performance tests.

Reply:
I've double check it on extended performance tests. It looks like there isn't neither improvement nor degradation of performance

<details>

<summary>Performance for SSE2 baseline</summary>

| Performance test | Reference time |  PR time | Speedup |
|------------------|----------------|----------|---------|
|pyrDown::Size_MatType::(127x61, 8UC1)|0.004|0.004|1.01|
|pyrDown::Size_MatType::(127x61, 16UC1)|0.006|0.006|1.01|
|pyrDown::Size_MatType::(127x61, 16SC1)|0.006|0.006|1.00|
|pyrDown::Size_MatType::(127x61, 32FC1)|0.004|0.004|1.01|
|pyrDown::Size_MatType::(127x61, 8UC3)|0.013|0.013|0.99|
|pyrDown::Size_MatType::(127x61, 16UC3)|0.014|0.014|1.01|
|pyrDown::Size_MatType::(127x61, 16SC3)|0.014|0.014|1.01|
|pyrDown::Size_MatType::(127x61, 32FC3)|0.014|0.014|1.00|
|pyrDown::Size_MatType::(127x61, 8UC4)|0.017|0.017|1.00|
|pyrDown::Size_MatType::(127x61, 16UC4)|0.013|0.013|1.02|
|pyrDown::Size_MatType::(127x61, 16SC4)|0.012|0.012|1.00|
|pyrDown::Size_MatType::(127x61, 32FC4)|0.018|0.018|1.00|
|pyrDown::Size_MatType::(320x240, 8UC1)|0.030|0.030|1.00|
|pyrDown::Size_MatType::(320x240, 16UC1)|0.047|0.047|1.00|
|pyrDown::Size_MatType::(320x240, 16SC1)|0.046|0.046|1.00|
|pyrDown::Size_MatType::(320x240, 32FC1)|0.026|0.025|1.03|
|pyrDown::Size_MatType::(320x240, 8UC3)|0.116|0.117|0.99|
|pyrDown::Size_MatType::(320x240, 16UC3)|0.123|0.122|1.01|
|pyrDown::Size_MatType::(320x240, 16SC3)|0.119|0.118|1.01|
|pyrDown::Size_MatType::(320x240, 32FC3)|0.122|0.125|0.98|
|pyrDown::Size_MatType::(320x240, 8UC4)|0.148|0.149|0.99|
|pyrDown::Size_MatType::(320x240, 16UC4)|0.107|0.105|1.02|
|pyrDown::Size_MatType::(320x240, 16SC4)|0.100|0.099|1.00|
|pyrDown::Size_MatType::(320x240, 32FC4)|0.160|0.163|0.98|
|pyrDown::Size_MatType::(640x480, 8UC1)|0.109|0.110|0.99|
|pyrDown::Size_MatType::(640x480, 16UC1)|0.180|0.180|1.00|
|pyrDown::Size_MatType::(640x480, 16SC1)|0.172|0.174|0.99|
|pyrDown::Size_MatType::(640x480, 32FC1)|0.096|0.097|0.99|
|pyrDown::Size_MatType::(640x480, 8UC3)|0.440|0.450|0.98|
|pyrDown::Size_MatType::(640x480, 16UC3)|0.466|0.477|0.98|
|pyrDown::Size_MatType::(640x480, 16SC3)|0.445|0.456|0.98|
|pyrDown::Size_MatType::(640x480, 32FC3)|0.468|0.482|0.97|
|pyrDown::Size_MatType::(640x480, 8UC4)|0.565|0.576|0.98|
|pyrDown::Size_MatType::(640x480, 16UC4)|0.399|0.406|0.98|
|pyrDown::Size_MatType::(640x480, 16SC4)|0.373|0.384|0.97|
|pyrDown::Size_MatType::(640x480, 32FC4)|0.627|0.647|0.97|
|pyrDown::Size_MatType::(1280x720, 8UC1)|0.309|0.309|1.00|
|pyrDown::Size_MatType::(1280x720, 16UC1)|0.526|0.529|0.99|
|pyrDown::Size_MatType::(1280x720, 16SC1)|0.507|0.515|0.98|
|pyrDown::Size_MatType::(1280x720, 32FC1)|0.270|0.273|0.99|
|pyrDown::Size_MatType::(1280x720, 8UC3)|1.336|1.331|1.00|
|pyrDown::Size_MatType::(1280x720, 16UC3)|1.380|1.410|0.98|
|pyrDown::Size_MatType::(1280x720, 16SC3)|1.324|1.362|0.97|
|pyrDown::Size_MatType::(1280x720, 32FC3)|1.531|1.611|0.95|
|pyrDown::Size_MatType::(1280x720, 8UC4)|1.720|1.736|0.99|
|pyrDown::Size_MatType::(1280x720, 16UC4)|1.206|1.232|0.98|
|pyrDown::Size_MatType::(1280x720, 16SC4)|1.148|1.170|0.98|
|pyrDown::Size_MatType::(1280x720, 32FC4)|2.116|2.165|0.98|
|pyrDown::Size_MatType::(1920x1080, 8UC1)|0.691|0.698|0.99|
|pyrDown::Size_MatType::(1920x1080, 16UC1)|1.174|1.198|0.98|
|pyrDown::Size_MatType::(1920x1080, 16SC1)|1.126|1.156|0.97|
|pyrDown::Size_MatType::(1920x1080, 32FC1)|0.677|0.692|0.98|
|pyrDown::Size_MatType::(1920x1080, 8UC3)|2.989|3.066|0.97|
|pyrDown::Size_MatType::(1920x1080, 16UC3)|3.170|3.225|0.98|
|pyrDown::Size_MatType::(1920x1080, 16SC3)|3.050|3.116|0.98|
|pyrDown::Size_MatType::(1920x1080, 32FC3)|3.647|3.629|1.01|
|pyrDown::Size_MatType::(1920x1080, 8UC4)|3.862|3.960|0.98|
|pyrDown::Size_MatType::(1920x1080, 16UC4)|2.872|2.947|0.97|
|pyrDown::Size_MatType::(1920x1080, 16SC4)|2.723|2.824|0.96|
|pyrDown::Size_MatType::(1920x1080, 32FC4)|4.881|4.841|1.01|
|pyrDown_ovx::Size_MatType::(127x61, 8UC1)|0.004|0.004|0.99|
|pyrDown_ovx::Size_MatType::(127x61, 16UC1)|0.006|0.006|1.01|
|pyrDown_ovx::Size_MatType::(127x61, 16SC1)|0.006|0.006|0.99|
|pyrDown_ovx::Size_MatType::(127x61, 32FC1)|0.004|0.004|1.04|
|pyrDown_ovx::Size_MatType::(127x61, 8UC3)|0.013|0.014|0.98|
|pyrDown_ovx::Size_MatType::(127x61, 16UC3)|0.014|0.014|1.01|
|pyrDown_ovx::Size_MatType::(127x61, 16SC3)|0.014|0.014|1.01|
|pyrDown_ovx::Size_MatType::(127x61, 32FC3)|0.014|0.014|1.00|
|pyrDown_ovx::Size_MatType::(127x61, 8UC4)|0.017|0.017|1.00|
|pyrDown_ovx::Size_MatType::(127x61, 16UC4)|0.013|0.013|1.01|
|pyrDown_ovx::Size_MatType::(127x61, 16SC4)|0.012|0.012|1.00|
|pyrDown_ovx::Size_MatType::(127x61, 32FC4)|0.018|0.018|1.00|
|pyrDown_ovx::Size_MatType::(320x240, 8UC1)|0.030|0.030|0.99|
|pyrDown_ovx::Size_MatType::(320x240, 16UC1)|0.047|0.047|1.01|
|pyrDown_ovx::Size_MatType::(320x240, 16SC1)|0.044|0.045|0.98|
|pyrDown_ovx::Size_MatType::(320x240, 32FC1)|0.025|0.025|0.99|
|pyrDown_ovx::Size_MatType::(320x240, 8UC3)|0.116|0.116|1.00|
|pyrDown_ovx::Size_MatType::(320x240, 16UC3)|0.121|0.120|1.01|
|pyrDown_ovx::Size_MatType::(320x240, 16SC3)|0.117|0.118|0.99|
|pyrDown_ovx::Size_MatType::(320x240, 32FC3)|0.123|0.123|1.00|
|pyrDown_ovx::Size_MatType::(320x240, 8UC4)|0.147|0.149|0.99|
|pyrDown_ovx::Size_MatType::(320x240, 16UC4)|0.104|0.105|0.99|
|pyrDown_ovx::Size_MatType::(320x240, 16SC4)|0.095|0.099|0.96|
|pyrDown_ovx::Size_MatType::(320x240, 32FC4)|0.158|0.163|0.97|
|pyrDown_ovx::Size_MatType::(640x480, 8UC1)|0.109|0.111|0.99|
|pyrDown_ovx::Size_MatType::(640x480, 16UC1)|0.170|0.181|0.94|
|pyrDown_ovx::Size_MatType::(640x480, 16SC1)|0.164|0.173|0.95|
|pyrDown_ovx::Size_MatType::(640x480, 32FC1)|0.091|0.095|0.95|
|pyrDown_ovx::Size_MatType::(640x480, 8UC3)|0.425|0.442|0.96|
|pyrDown_ovx::Size_MatType::(640x480, 16UC3)|0.449|0.473|0.95|
|pyrDown_ovx::Size_MatType::(640x480, 16SC3)|0.423|0.466|0.91|
|pyrDown_ovx::Size_MatType::(640x480, 32FC3)|0.445|0.468|0.95|
|pyrDown_ovx::Size_MatType::(640x480, 8UC4)|0.542|0.565|0.96|
|pyrDown_ovx::Size_MatType::(640x480, 16UC4)|0.380|0.398|0.95|
|pyrDown_ovx::Size_MatType::(640x480, 16SC4)|0.363|0.373|0.97|
|pyrDown_ovx::Size_MatType::(640x480, 32FC4)|0.596|0.627|0.95|
|pyrDown_ovx::Size_MatType::(1280x720, 8UC1)|0.309|0.309|1.00|
|pyrDown_ovx::Size_MatType::(1280x720, 16UC1)|0.502|0.518|0.97|
|pyrDown_ovx::Size_MatType::(1280x720, 16SC1)|0.502|0.505|0.99|
|pyrDown_ovx::Size_MatType::(1280x720, 32FC1)|0.256|0.264|0.97|
|pyrDown_ovx::Size_MatType::(1280x720, 8UC3)|1.334|1.351|0.99|
|pyrDown_ovx::Size_MatType::(1280x720, 16UC3)|1.314|1.372|0.96|
|pyrDown_ovx::Size_MatType::(1280x720, 16SC3)|1.271|1.325|0.96|
|pyrDown_ovx::Size_MatType::(1280x720, 32FC3)|1.445|1.510|0.96|
|pyrDown_ovx::Size_MatType::(1280x720, 8UC4)|1.638|1.721|0.95|
|pyrDown_ovx::Size_MatType::(1280x720, 16UC4)|1.149|1.206|0.95|
|pyrDown_ovx::Size_MatType::(1280x720, 16SC4)|1.091|1.143|0.95|
|pyrDown_ovx::Size_MatType::(1280x720, 32FC4)|2.037|2.108|0.97|
|pyrDown_ovx::Size_MatType::(1920x1080, 8UC1)|0.685|0.694|0.99|
|pyrDown_ovx::Size_MatType::(1920x1080, 16UC1)|1.154|1.192|0.97|
|pyrDown_ovx::Size_MatType::(1920x1080, 16SC1)|1.125|1.152|0.98|
|pyrDown_ovx::Size_MatType::(1920x1080, 32FC1)|0.679|0.665|1.02|
|pyrDown_ovx::Size_MatType::(1920x1080, 8UC3)|2.985|3.064|0.97|
|pyrDown_ovx::Size_MatType::(1920x1080, 16UC3)|3.178|3.225|0.99|
|pyrDown_ovx::Size_MatType::(1920x1080, 16SC3)|3.048|3.123|0.98|
|pyrDown_ovx::Size_MatType::(1920x1080, 32FC3)|3.602|3.613|1.00|
|pyrDown_ovx::Size_MatType::(1920x1080, 8UC4)|3.865|3.960|0.98|
|pyrDown_ovx::Size_MatType::(1920x1080, 16UC4)|2.899|2.866|1.01|
|pyrDown_ovx::Size_MatType::(1920x1080, 16SC4)|2.749|2.798|0.98|
|pyrDown_ovx::Size_MatType::(1920x1080, 32FC4)|4.873|4.885|1.00|
|pyrUp::Size_MatType::(127x61, 8UC1)|0.014|0.014|1.00|
|pyrUp::Size_MatType::(127x61, 16UC1)|0.017|0.017|0.97|
|pyrUp::Size_MatType::(127x61, 16SC1)|0.016|0.016|1.00|
|pyrUp::Size_MatType::(127x61, 32FC1)|0.016|0.015|1.01|
|pyrUp::Size_MatType::(127x61, 8UC3)|0.038|0.038|1.00|
|pyrUp::Size_MatType::(127x61, 16UC3)|0.047|0.049|0.96|
|pyrUp::Size_MatType::(127x61, 16SC3)|0.045|0.045|1.00|
|pyrUp::Size_MatType::(127x61, 32FC3)|0.044|0.044|1.00|
|pyrUp::Size_MatType::(127x61, 8UC4)|0.048|0.049|1.00|
|pyrUp::Size_MatType::(127x61, 16UC4)|0.062|0.065|0.96|
|pyrUp::Size_MatType::(127x61, 16SC4)|0.059|0.059|1.00|
|pyrUp::Size_MatType::(127x61, 32FC4)|0.059|0.060|0.98|
|pyrUp::Size_MatType::(320x240, 8UC1)|0.117|0.116|1.01|
|pyrUp::Size_MatType::(320x240, 16UC1)|0.152|0.157|0.97|
|pyrUp::Size_MatType::(320x240, 16SC1)|0.143|0.142|1.00|
|pyrUp::Size_MatType::(320x240, 32FC1)|0.145|0.145|1.00|
|pyrUp::Size_MatType::(320x240, 8UC3)|0.342|0.345|0.99|
|pyrUp::Size_MatType::(320x240, 16UC3)|0.446|0.458|0.97|
|pyrUp::Size_MatType::(320x240, 16SC3)|0.412|0.414|1.00|
|pyrUp::Size_MatType::(320x240, 32FC3)|0.456|0.448|1.02|
|pyrUp::Size_MatType::(320x240, 8UC4)|0.461|0.465|0.99|
|pyrUp::Size_MatType::(320x240, 16UC4)|0.600|0.608|0.99|
|pyrUp::Size_MatType::(320x240, 16SC4)|0.554|0.547|1.01|
|pyrUp::Size_MatType::(320x240, 32FC4)|0.598|0.589|1.02|
|pyrUp::Size_MatType::(640x480, 8UC1)|0.453|0.463|0.98|
|pyrUp::Size_MatType::(640x480, 16UC1)|0.591|0.612|0.96|
|pyrUp::Size_MatType::(640x480, 16SC1)|0.551|0.551|1.00|
|pyrUp::Size_MatType::(640x480, 32FC1)|0.581|0.572|1.02|
|pyrUp::Size_MatType::(640x480, 8UC3)|1.364|1.422|0.96|
|pyrUp::Size_MatType::(640x480, 16UC3)|1.742|1.813|0.96|
|pyrUp::Size_MatType::(640x480, 16SC3)|1.634|1.641|1.00|
|pyrUp::Size_MatType::(640x480, 32FC3)|1.951|1.940|1.01|
|pyrUp::Size_MatType::(640x480, 8UC4)|1.891|1.933|0.98|
|pyrUp::Size_MatType::(640x480, 16UC4)|2.361|2.441|0.97|
|pyrUp::Size_MatType::(640x480, 16SC4)|2.199|2.197|1.00|
|pyrUp::Size_MatType::(640x480, 32FC4)|2.790|2.750|1.01|
|pyrUp::Size_MatType::(1280x720, 8UC1)|1.393|1.402|0.99|
|pyrUp::Size_MatType::(1280x720, 16UC1)|1.791|1.842|0.97|
|pyrUp::Size_MatType::(1280x720, 16SC1)|1.658|1.676|0.99|
|pyrUp::Size_MatType::(1280x720, 32FC1)|1.977|1.944|1.02|
|pyrUp::Size_MatType::(1280x720, 8UC3)|4.122|4.188|0.98|
|pyrUp::Size_MatType::(1280x720, 16UC3)|5.290|5.474|0.97|
|pyrUp::Size_MatType::(1280x720, 16SC3)|5.086|5.096|1.00|
|pyrUp::Size_MatType::(1280x720, 32FC3)|6.022|6.051|1.00|
|pyrUp::Size_MatType::(1280x720, 8UC4)|5.683|5.526|1.03|
|pyrUp::Size_MatType::(1280x720, 16UC4)|7.177|7.420|0.97|
|pyrUp::Size_MatType::(1280x720, 16SC4)|6.825|6.813|1.00|
|pyrUp::Size_MatType::(1280x720, 32FC4)|8.813|8.782|1.00|


</details>